doc/contrib/copy_on_write.md

```diff
@@ -0,0 +1,85 @@
+# Copy-on-Write
+
+The two basic Elektra datastructures `Key` and `KeySet` implement full copy-on-write (COW) functionality.
+If a key or a keyset gets copied, only a shallow copy with references to the original name and value is created.
+Only if it gets modified the necessary memory is actually allocated.
+As a consequence, duplicated keys or keysets only require a fraction of the memory compared to their source counterparts.
+
+## Working principle
+
+The main principle for the copy-on-write approach is simple: seperate the "real" data from the user-facing identifying entity, in our case `Key` and `KeySet`.
+Developers using Elektra only get references to `Key` and `KeySet` objects.
+Those objects themselves do not hold any data.
+They only contain references to other entities holding the data.
+
+The main entities for copy-on-write are depicted in the following diagram:
+
+```mermaid
+classDiagram
+    Key *-- "0..1" KeyName : keyName
+    Key *-- "0..1" KeyData : keyData
+
+    KeySet *-- "0..1" KeySetData : data
+
+    Key o-- "0..1" KeySet : meta
+    KeySetData o-- "*" Key : array
+
+    Key : - uint16_t refs
+    Key : - keyflag_t flags
+
+    KeyName : - uint16_t refs
+    KeyName : - uint16_t flags
+
+    KeyData : - uint16_t refs
+    KeyData : - uint16_t flags
+
+    KeySet : - uint16_t refs
+    KeySet : - ksflag_t flags
+
+    KeySetData : - uint16_t refs
+    KeySetData : - uint16_t flags
+```
+
+- `Key`: logically represents a key. No real data is actually stored in it.
+  - `KeyName`: holds the name of a key. May be shared between multiple `Key` objects.
+  - `KeyData`: holds the value of a key. May be shared between multiple `Key` objects.
+- `KeySet`: logically represents a collection of keys. No real data is actually stored in it.
+  - `KeySetData`: holds an array of `Key` objects. May be shared between multiple `KeySet` objects.
+
+When creating instances of `Key` and `KeySet` via `keyNew` or `ksNew`, we allocate the least amount.
+For `Key`, we only allocate a `KeyName` entity.
+For `KeySet`, we allocate no other entities.
+The missing entities will get allocated when needed.
+
+If the name of a key gets modified, we first ensure that the key is the singular "owner" of its `KeyName` entity.
```

Comment:
Please use terminology properly (somewhere write what owner is, what owns mean etc.). We can use owner explicitly for COW, it is currently not used anywhere else (we had another owner concept in the past, but this, except of small left-overs waiting to be removed, not present anymore).

```suggestion
If the name of a key gets modified, we first ensure that the key is the single owner of its `KeyName` entity.
```

Reply:
This has already been changed. I think you commented on a pretty outdated version of the documentation.